src/chat/tools/applyTemplateTool.ts

```diff
@@ -0,0 +1,173 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import * as vscode from "vscode";
+import {
+  CancellationToken,
+  ChatRequest,
+  ChatResponseStream,
+  LanguageModelChatMessage,
+  LanguageModelTextPart,
+  LanguageModelToolCallPart,
+  LanguageModelToolInvocationOptions,
+  LanguageModelToolResult,
+} from "vscode";
+import { Logger } from "../../logging";
+import { scaffoldProjectRequest } from "../../scaffold";
+import { logUsage, UserEvent } from "../../telemetry/events";
+import { BaseLanguageModelTool } from "./base";
+
+const logger = new Logger("chat.tools.applyTemplate");
+
+export interface IApplyTemplateParameters {
+  name: string;
+  options: { [key: string]: string };
+}
+
+/** Parse a LanguageModelTextPart from ListTemplatesTool into IApplyTemplateParameters. */
+export function parseListTemplatesOutput(part: LanguageModelTextPart): IApplyTemplateParameters {
+  const match = part.value.match(/id="(.+?)"; inputOptions=(\{.*\})/);
+  if (!match) {
+    throw new Error("Invalid template output format");
+  }
+
+  const name = match[1];
+  const options = match[2] ? JSON.parse(match[2]) : {};
+
+  return { name, options };
+}
+
+export class ApplyTemplateTool extends BaseLanguageModelTool<IApplyTemplateParameters> {
+  readonly name = "apply_projectTemplate";
+  readonly progressMessage = "Applying the selected project template...";
+
+  async prepareInvocation(
+    options: vscode.LanguageModelToolInvocationPrepareOptions<IApplyTemplateParameters>,
+    _token: vscode.CancellationToken,
+  ): Promise<vscode.PreparedToolInvocation | null | undefined> {
+    try {
+      const inputPart = new LanguageModelTextPart(
+        `id="${options.input.name}"; inputOptions=${JSON.stringify(options.input.options)}`,
+      );
+      const parsedParameters = parseListTemplatesOutput(inputPart);
+
+      options.input.options = { ...parsedParameters.options, ...options.input.options };
+    } catch (error) {
+      logger.error("Error parsing template output:", error);
+      throw new Error("Failed to parse template output.");
+    }
```

Comment:
As long as the `*api_key`/`*api_secret` options are removed from the list-templates tool response, we should be able to remove this (and `parseListTemplatesOutput`) entirely. If the model is behaving as expected, we should see it triggering a tool invocation for `GetProjectTemplateTool` and passing the options in here directly.